Bug 1223560: Use WhiteNoise to serve static assets

openjck opened the pull request:
<!-- Reviewable:start -->
[<img src="https://reviewable.io/review_button.png" height=40 alt="Review on Reviewable"/>](https://reviewable.io/reviews/mozilla/kuma/3642)
<!-- Reviewable:end -->

openjck commented on provisioning/roles/kuma/files/apache2/all-servers.conf:
```diff
@@ -3,15 +3,6 @@
-    SetEnvIf Origin "https?://(.*\.(mozilla|allizom)\.(com|org|net))" CORS=$0
-    SetEnvIf Origin "https?://(mdn\.mozillademos\.org)" CORS=$0
-    SetEnvIf Origin "https?://(kuma\.readthedocs\.org)" CORS=$0
-    Header set Access-Control-Allow-Origin %{CORS}e env=CORS
```
@groovecoder Do I remember correctly that we decided not to worry about restricting font CORS for approved any more? I'm pretty sure we did, but somehow these lines stuck around for some reason.

openjck replied:
Hm... I guess I'm not sure about whether we want to remove this. I can restore this locally and verify that it still works once I'm able to test.

jezdez replied:
Note, Whitnoise will add `Access-Control-Allow-Origin=*` to all files it serves by default. If you think we should disable it, you can use the `WHITENOISE_ALLOW_ALL_ORIGINS` setting (http://whitenoise.readthedocs.org/en/latest/django.html?highlight=WHITENOISE_ALLOW_ALL_ORIGINS#WHITENOISE_ALLOW_ALL_ORIGINS) or pass `allow_all_origins=False` to the instantiation of `KumaDjangoWhiteNoise` in kuma.wsgi.

jezdez replied:
Here's some more explanation why it does that: https://github.com/evansd/whitenoise/blob/0657e7aea986fa652bd04d6ae8468a0e1129d634/whitenoise/base.py#L90-L95

openjck replied:
Great. Thanks for pointing me to that documentation.

My understanding is that these lines were added to prevent others from hotlinking licensed fonts. There doesn't seem to be any documentation supporting that, though, so I'll assume it's safe to disable that protection unless someone knows otherwise.

openjck commented on static/.htaccess:
```diff
@@ -1,29 +1 @@
-ExpiresActive on
 FileETag MTime
```
@jezdez Do we still need this or will Whitenoise and/or Django take care of it from now on?

jezdez replied:
Whitenoise doesn't provide Etag support out of the box since it's by design something that is left to the server to implement. The format of Etag heavily differ between server software, given our jump away from Apache soon as part of the AWS move (going to Nginx) we'll need to reevaluate the use of Etag's anyway. I'd say we remove them and only live with the ones that Whitenoise sets: `Last-Modified`, `Content-Length`, `Content-Type`, `Content-Encoding`, `Cache-Control`. That should give our CDN enough info to decide whether to fetch the files or not.

jezdez commented on wsgi/kuma.wsgi:
```diff
@@ -29,6 +29,10 @@ import manage
 
+# Enable WhiteNoise
+from whitenoise.django import DjangoWhiteNoise
+django_app = DjangoWhiteNoise(django_app)
```
There are  a few options we can set here, e.g. subclassing the class and modifying the `EXTRA_MIMETYPES` value may make sense to support more of those we have in the Apache config: https://github.com/evansd/whitenoise/blob/b3974c2ea80e93ba086331cff447a09a427a1a9f/whitenoise/base.py#L72-L73

openjck commented:
Updated!

I couldn't find an easy way to test this locally. Is there something you recommend?

robhudson commented on kuma/core/whitenoise/KumaDjangoWhiteNoise.py:
```diff
@@ -0,0 +1,12 @@
+class KumaDjangoWhiteNoise(DjangowhiteNoise):
+
+    def __init__(self, app):
+        DjangoWhiteNoise.__init__(self, app)
```
What about `super(DjangoWhiteNoise, self).__init__(app)`? But also if you're not adding anything here there's no need, Python will do this for you.

openjck replied:
Oh okay, great. I'll remove this then.

robhudson commented on kuma/core/whitenoise/KumaDjangoWhiteNoise.py:
```diff
@@ -0,0 +1,12 @@
+    # Correct some MIME types
+    EXTRA_MIMETYPES = (
+            ('application/font-woff', '.woff'),
+            ('application/font-woff2', '.woff2'),)
```
Should we create this like how it is done in the tests?
https://github.com/evansd/whitenoise/blob/bdeea1a18b134ba59a9f6a31dd2f1839836b4b14/tests/test_whitenoise.py#L42-L43

openjck replied:
In terms of adding to the existing tuple, rather than replacing it? I started down that path, but then realized that the default `woff` MIME type that WhiteNoise uses is wrong. I'll have to open an issue with them about it.

https://github.com/evansd/whitenoise/blob/b3974c2ea80e93ba086331cff447a09a427a1a9f/whitenoise/base.py#L72-L73

openjck replied:
But then, maybe a better approach would be to replace the inner tuple relating to `woff2`, so that we're completely overriding `EXTRA_MIMETYPES` given that it might contain some useful defaults later on. I'll give it a go.

robhudson replied:
Ah ok, I didn't see that they were also setting this and that this value was so sparse. I imagined it having a lot more. This seems fine but perhaps with a caveat if whitenoise adds more in the future?

openjck commented on kuma/core/whitenoise/KumaDjangoWhiteNoise.py:
```diff
@@ -0,0 +1,16 @@
+    # defaults), use ours.
+    EXTRA_MIMETYPES = tuple({e[1]: e for e in reversed(
+            EXTRA_MIMETYPES_KUMA + DjangoWhiteNoise.EXTRA_MIMETYPES
+        )}.values())
```
I'm sure there there's a better way to do this. But this works, and it gives a sense of what I'm going for.

openjck commented on kuma/core/whitenoise/KumaDjangoWhiteNoise.py:
```diff
@@ -0,0 +1,15 @@
+    # WhiteNoise. When our MIME types conflict with the defaults, use ours.
+    EXTRA_MIMETYPES = tuple({e[1]: e for e in reversed(
+            KUMA_MIMETYPES + DjangoWhiteNoise.EXTRA_MIMETYPES
+        )}.values())
```
I'm sure there there's a better way to do this. But it works, and it illustrates what I'm going for.

openjck commented:
Tests are passing. This is ready for review. :smile:

jezdez commented on kuma/core/whitenoise/KumaDjangoWhiteNoise.py:
```diff
@@ -0,0 +1,15 @@
+from whitenoise.django import DjangoWhiteNoise
```
Please rename the file to `kuma/core/whitenoise.py` and then add the following as the first line to that file so it doesn't end up with a circular import error due to the equally named local module `whitenoise.py` and the package called `whitenoise`:

```python
from __future__ import absolute_import
```

That makes the feature that is in Python 3.x work on Python 2.x that we run under. More info here: https://www.python.org/dev/peps/pep-0328/

jezdez commented on kuma/core/whitenoise/KumaDjangoWhiteNoise.py:
```diff
@@ -0,0 +1,15 @@
+    # Custom MIME types
+    KUMA_MIMETYPES = (
+        ('application/font-woff', '.woff'),
+        ('application/font-woff2', '.woff2'),)
```
Whitenoise uses `font/woff2` as the mimetype for that file extension, that's no good?

jezdez commented on kuma/core/whitenoise/KumaDjangoWhiteNoise.py:
```diff
@@ -0,0 +1,15 @@
+    # WhiteNoise. When our MIME types conflict with the defaults, use ours.
+    EXTRA_MIMETYPES = tuple({e[1]: e for e in reversed(
+        KUMA_MIMETYPES + DjangoWhiteNoise.EXTRA_MIMETYPES
+    )}.values())
```
Ough, yeah, I'd rather just redefine the tuple completely in this case. This magic is not worth the effort given the fact that the upstream whitenoise class currently only has one item: https://github.com/evansd/whitenoise/blob/bdeea1a18b134ba59a9f6a31dd2f1839836b4b14/whitenoise/base.py#L72-L73

openjck replied:
Yeah, there's only one item in the parent class at the moment. Would this have a noticeable performance impact given that we're using a CDN? WhiteNoise could add more items in the future and this would be easier than watching their code for important changes.

jezdez commented on static/.htaccess:
```diff
@@ -1,16 +0,0 @@
-ExpiresByType video/webm "access plus 1 week"
-ExpiresByType video/ogg "access plus 1 week"
-ExpiresByType video/x-flv "access plus 1 week"
-ExpiresByType application/x-shockwave-flash "access plus 1 week"
```
I was wondering if we should work on making the expiration differ per file type as well, I think that'd be possible with the hooks in whitenoise. What do you think?

openjck replied:
Glad you brought this up. I wasn't sure about this. The WhiteNoise docs say the following about expiration values:

> The default is chosen to be short enough not to cause problems with stale versions but long enough that, if you’re running WhiteNoise behind a CDN, the CDN will still take the majority of the strain during times of heavy load.

Does that mean WhiteNoise will choose some reasonable defaults for us?

jezdez replied:
Whitenoise will by default use an expiration time of 60 seconds for non-hashed files: https://github.com/evansd/whitenoise/blob/0657e7aea986fa652bd04d6ae8468a0e1129d634/whitenoise/base.py#L89

For hashed files ("immutable files") it'll use a far future value: https://github.com/evansd/whitenoise/blob/0657e7aea986fa652bd04d6ae8468a0e1129d634/whitenoise/base.py#L229-L230

So I think we should be good since in production we'll ever only use hashed values and for the non-hashed files we can allow the CDN to fetch the files every 60 seconds IMO.

openjck replied:
Agreed. To my knowledge, the only pages referencing non-hashed assets are the pages of our Read the Docs theme. We can even bump the default value to "access + 1 week" since that's the lowest value we use now.

jezdez commented on wsgi/kuma.wsgi:
```diff
@@ -29,6 +29,10 @@ import manage
 django_app = get_wsgi_application()
 
+# Enable WhiteNoise
+from kuma.core.whitenoise import KumaDjangoWhiteNoise
```
I think this should have been `from kuma.core.whitenoise.KumaDjangoWhiteNoise import KumaDjangoWhiteNoise` for the case in which the file is named `KumaDjangoWhiteNoise.py`. But if you change the `KumaDjangoWhiteNoise` class to be contained in `kuma.core.whitenoise` then this is correct.

openjck replied:
Great! That was an oversight alright, but it works with the new path.

jezdez commented:
Only spotchecked so far, haven't tried in a VM yet since that's broken for me right now :(

openjck commented:
Updated. I haven't been able to spot-check this yet. Is there any easy way to run Kuma locally with wsgi?

openjck commented:
Hm... here's another idea. We can use Pipeline to set MIME types and avoid subclassing `DjangoWhiteNoise` altogether.

https://django-pipeline.readthedocs.org/en/latest/configuration.html#pipeline-mimetypes

I'll move ahead with that plan unless there are any objections.

openjck commented:
Oh nevermind, I guess that won't work because Pipeline isn't serving the assets on prod.

jezdez commented on kuma/core/whitenoise.py:
```diff
@@ -0,0 +1,15 @@
+    # WhiteNoise. When our MIME types conflict with the defaults, use ours.
+    EXTRA_MIMETYPES = tuple({e[1]: e for e in reversed(
+        KUMA_MIMETYPES + DjangoWhiteNoise.EXTRA_MIMETYPES
+    )}.values())
```
This class attribute will only be evaluated once, but it's kind of smelly code and to be honest I think we should not accidentally change the mimetypes we support when we update whitenoise and it adds a new entry. In other words, let's make this a static list of mimetypes we know we want to support and ignore the upstream mimetypes.

jezdez commented:
@openjck Since we currently still haven't updated key aspects of the way how we run the Python code we can't just point a Python server to pick up the wsgi file. Sorry.

openjck commented:
Updated!

I added the *kuma.wsgi* file you sent me yesterday and updated *Procfile* to use it. I know that's not what we talked about earlier, but it looks like we need that to properly test assets moving forward.

jezdez commented on kuma/vagrant-wsgi.py:
```diff
@@ -0,0 +1,14 @@
+import os
```
Sorry, this file should be called `kuma/wsgi.py`.

openjck replied:
Sure, can do. Thanks for the feedback.

Is there a name we could use that would differentiate this from *wsgi/kuma.py*? The names are close enough to be confusing. Or is the goal to incorporate *wsgi/kuma.py* into this file over time?

jezdez replied:
The goal (once we're on AWS) is to not have a `wsgi/kuma.wsgi` anymore as that's an artifact of the way we currently run MDN via Apache w/ mod_wsgi. On AWS we'll be using Nginx and uwsgi.

jezdez commented on provisioning/roles.txt:
```diff
@@ -5,3 +5,4 @@ geerlingguy.mysql,1.5.0
 Mayeu.RabbitMQ,1.4.0
 nodesource.node
 Stouts.elasticsearch,2.5.0
+gunicorn,19.3
```
This isn't the right file for Python dependencies. Add this to the file `/provisioning/roles/kuma/tasks/python.yml` after the installation of `performance.txt` requirements file:

```yml
- name: Install some more deps in the virtualenv
  pip:
    virtualenv: "/home/vagrant/env"
    virtualenv_site_packages: yes
    name: "{{ item }}"
  with_items:
    - gunicorn
```

jezdez commented on Procfile:
```diff
@@ -1,4 +1,4 @@
-web: make clean && python2.7 manage.py runserver 0.0.0.0:8000
+web: make clean && gunicorn kuma.vagrant-wsgi -b 0.0.0.0:8000
```
That means we'll stop using the django dev server which serves static files using the staticfiles system and instead uses whitenoise in development. I wonder if that means we'll have to run collectstatic before we run it now. Can you double check if a changed file shows up on the site without restarting Foreman?

openjck replied:
Good catch. It works the same way as `runserver` for me&mdash;collectstatic only needs to be run when `DEBUG = True`.

jezdez commented on provisioning/roles/kuma/files/vagrant/settings_local.py:
```diff
@@ -108,3 +108,7 @@
+
+# Don't cache non-versioned static files in DEBUG mode
+if DEBUG:
+    WHITENOISE_MAX_AGE = 0
```
I think that's the default, https://github.com/evansd/whitenoise/blob/0657e7aea986fa652bd04d6ae8468a0e1129d634/whitenoise/django.py#L58-L59 Unless you want to allow people to change it?

openjck replied:
We [override the default](https://github.com/mozilla/kuma/pull/3642/files#diff-bdf3ecebd8379ca98cc89e545fc90899R557) in *settings.py*. These lines override that value when `DEBUG = True`.

jezdez commented on settings.py:
```diff
@@ -553,6 +553,9 @@ def JINJA_CONFIG():
 STATICI18N_DOMAIN = 'javascript'
 
+# Cache non-versioned static files for one week
+WHITENOISE_MAX_AGE = 604800
```
Would you mind writing that in the form of `60 * 60 * 24 * 7`? That has in my experience removed the magic behind such numbers and prevented confusion when others start working on the project.

openjck replied:
Great, completely agree. Will do.

openjck commented:
All good suggestions. Updated! :frog:

jezdez commented on requirements/source.txt:
```diff
@@ -115,3 +115,5 @@ django-sundial==1.0.1
 
 django-celery-transactions==0.3.6
+
+whitenoise==2.0.4
```
There is a newer 2.0.6 release, sorry :(

jezdez commented on settings.py:
```diff
@@ -559,6 +559,9 @@ def JINJA_CONFIG():
 
+# Cache non-versioned static files for one week
+WHITENOISE_MAX_AGE = 60 * 60 * 24 * 7
+
```
\o/

jezdez commented:
LGTM r+
![fpjcjvuv1h1pa](https://cloud.githubusercontent.com/assets/1610/11575597/2ee3b068-9a12-11e5-8f94-480b3c8447e5.gif)

openjck commented:
Updated! The latest version of WhiteNoise uses exactly the `EXTRA_MIMETYPES` we need, so I removed our subclass. :raised_hands:

jezdez commented:
@openjck Nice!

jezdez commented:
r+

openjck commented:
Great!

This needs some coordination with WebOps. Basically, we'll need to have them stop hosting the *static* directory with Apache so that Python can do the work instead.

Let's merge this early next week and take it from there.

openjck commented:
Tagging as not ready (to be merged) because we want to coordinate this merge/push with WebOps tomorrow.
